src/background/sidebar-injector.js

```diff
@@ -336,6 +350,81 @@ export function SidebarInjector() {
     return chromeAPI.tabs.executeScript(tab.id, { file: '/unload-client.js' });
   }
 
+  /**
+   * Find the frame within the VitalSource Bookshelf reader into which the
+   * Hypothesis client should be loaded.
+   *
+   * The frame hierarchy will look like:
+   *
+   * bookshelf.vitalsource.com (Main frame)
+   * |- jigsaw.vitalsource.com (Ebook reader. This is where we want to inject the client)
+   *     |- jigsaw.vitalsource.com (Content of current chapter)
+   *
+   * @param {Tab} tab
+   */
+  async function getVitalSourceViewerFrame(tab) {
+    // Using `chrome.webNavigation.getAllFrames` requires asking for the
+    // `webNavigation` permission which results in a scary prompt about reading
+    // browser history, even though we only want to get frames for the current
+    // tab :(
+    //
+    // A more roundabout way to solve this might be to use `executeScript({ allFrames: true, ... })`
+    // to discover frame IDs and URLs of each frame in the tab.
+    const canUseWebNavigation = await chromeAPI.permissions.request({
+      permissions: ['webNavigation'],
+    });
+    if (!canUseWebNavigation) {
+      throw new Error('The extension was not granted required permissions');
+    }
+
+    const frames = await chromeAPI.webNavigation.getAllFrames({
+      tabId: tab.id,
+    });
+    if (!frames) {
+      throw new Error('Could not list frames in tab');
+    }
+
+    return frames.find(frame => {
+      const frameURL = new URL(frame.url);
+      if (
+        frameURL.hostname !== 'jigsaw.vitalsource.com' ||
+        !frameURL.pathname.startsWith('/mosaic/wrapper.html')
+      ) {
```

Comment:
I sense a touch of long-term fragility here, but probably nothing you're not aware of already...